File: collect_app/src/test/java/org/odk/collect/android/support/InstanceUtils.java
package org.odk.collect.android.support;

import org.odk.collect.android.instances.Instance;

import java.io.File;

public class InstanceUtils {

    private InstanceUtils() {

    }

    public static Instance.Builder buildInstance(String formId, String version, String instancesDir) {
        return buildInstance(formId, version, "display name", Instance.STATUS_INCOMPLETE, null, instancesDir);
    }

    public static Instance.Builder buildInstance(String formId, String version, String displayName, String status, Long deletedDate, String instancesDir) {
        File instanceDir = new File(instancesDir + File.separator + System.currentTimeMillis() + Math.random());
        instanceDir.mkdir();

        return new Instance.Builder()
                .formId(formId)
                .formVersion(version)
                .displayName(displayName)
                .instanceFilePath(instanceDir.getAbsolutePath())
                .status(status)
                .lastStatusChangeDate(System.currentTimeMillis())
                .status(status)
                .deletedDate(deletedDate);
    }
}

File: forms-test/src/main/java/org/odk/collect/formstest/InstanceUtils.kt
package org.odk.collect.formstest

import org.odk.collect.forms.instances.Instance
import org.odk.collect.shared.TempFiles.createTempFile
import org.odk.collect.shared.strings.RandomString
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

object InstanceUtils {

    @JvmStatic
    fun buildInstance(formId: String?, version: String?, instancesDir: String): Instance.Builder {
        return buildInstance(
            formId,
            version,
            "display name",
            Instance.STATUS_INCOMPLETE,
            null,
            instancesDir
        )
    }

    @JvmStatic
    fun buildInstance(
        formId: String?,
        version: String?,
        displayName: String?,
        status: String?,
        deletedDate: Long?,
        instancesDir: String
    ): Instance.Builder {
        val instanceFile = createInstanceDirAndFile(instancesDir)

        return Instance.Builder()
            .formId(formId)
            .formVersion(version)
            .displayName(displayName)
            .instanceFilePath(instanceFile.absolutePath)
            .status(status)
            .deletedDate(deletedDate)
    }

    @JvmStatic
    fun createInstanceDirAndFile(instancesDir: String): File {
        val timestamp = SimpleDateFormat("yyyy-MM-dd_HH-mm-ss", Locale.ENGLISH)
            .format(Date((100_000_000_0000L..999_999_999_9999L).random()))

        val instanceDir = File(instancesDir + File.separator + RandomString.randomString(5) + "_" + timestamp)
        instanceDir.mkdir()

        return createTempFile(instanceDir, instanceDir.name, ".xml").also {
            it.writeText(RandomString.randomString(10))
        }
    }
}
